astropy/io/votable/tests/test_vo.py

```diff
@@ -670,18 +670,18 @@ def setup_class(self):
             votable = parse(get_pkg_data_filename("data/regression.xml"))
         votable.version = "1.3"
         votable.get_first_table()._config["version_1_3_or_later"] = True
-        votable.get_first_table().format = "binary2"
+        
 
         self.xmlout = bio = io.BytesIO()
         # W39: Bit values can not be masked
-        with pytest.warns(W39):
-            votable.to_xml(bio)
+        votable.to_xml(bio)
         bio.seek(0)
         self.votable = parse(bio)
-
+        self.votable.get_first_table().format = "binary2"
```

Comment:
This was a careless mistake on my part (will reset); I was confused as to why the format appeared as `tabledata` rather than `binary2`, so I assumed it had been overwritten somewhere (still wondering why it appears as `tabledata`, though seems unrelated), causing the test to fail. After reviewing the VOTable Format Definition for 1.3 again, I see that the reason this specific test was failing this test was actually because:

> "The only way to mark as null individual elements of an array-valued cell is by use of the **magic value mechanism**, which operates on a per-element basis. Although the magic value approach can mark individual elements of an array as null, it cannot mark a whole multi-element array as null."

https://www.ivoa.net/documents/VOTable/20130920/REC-VOTable-1.3-20130920.html#sec:NULL
The last sentence being one of the appeals of `BINARY2`.

This makes sense to me for what the original implementer was going for as:
```python
# This will simply mark the whole multi-element array as null if the array_mask at that index is specified as True entirely
>> flattened = np.array([np.all(x) for x in array_mask]) <<
data.write(converters.bool_to_bitarray(flattened))
```

And since the test is specifically regarding a multi-element array of booleans, this causes it to fail. On the other hand, a test such as `test_nulls_array` (requiring individual elements in the cell's array to be nulled) would pass as it _does_ make use of the magic value mechanism. This seems like a genuine mistake to reuse this test for `BINARY2` serialization, however, I would like to confirm with someone to be 100% before I make any further changes to this. 

Also, yes, I completely understand on the wait, I thought it would be a simple fix- I would like not to mess things up for future users 😆

Reply:
Thank you for your detailed and diligent response, it's really helpful !
It's great that you managed to pass existing tests with no modifications, now I think the priority is to add a new regression test that fails on main and passes with your patch.

Regarding the alleged[^1] genuine mistake in `test_nulls_array`, is it blocking your work or can it be addressed independently ? In the latter case, I suggest opening a separate issue for it so it can get the attention it needs.


[^1]: I'm not suggesting you're wrong, I'm just cautiously avoiding to express an opinion I don't have.

Reply:
@neutrinoceros I'm also on the fence about it being a mistake (so I'll also say perhaps allegedly too haha), but assuming they intended for the two masking mechanisms to be used (which I'm beginning to think that is the case, given that the tests were reused and perhaps why they may have initially allowed the masking to be handled twice that was their initial implementation: `BINARY2` handles what it can for masking and then to be sure also pass it through the converters for masking it can't handle), the new fix addresses it by specifically including in the condition what the `BINARY2` masking cannot handle (array elements) and doesn't block my work (will add a new regression test very soon)

In a nutshell, quick summary (astropy's original code):
https://github.com/astropy/astropy/blob/0f1e141f2e9ecf48be9e1465caf9f9d4b8464bc7/astropy/io/votable/tree.py#L3295-L3301
-> Applies BINARY2 masking.
-> Also, uses a general masking approach to cover scenarios that `BINARY2` masking might miss (converters)
-> But **will** also mask the ones `BINARY2` **_can_** handle (causing issue #16090)

The fix:
https://github.com/astropy/astropy/blob/a378244646dab7d8ccdb28f9e132e0e2fe626c40/astropy/io/votable/tree.py#L3303-L3310
-> Applies BINARY2 masking.
-> Also, uses a general masking approach to cover scenarios that `BINARY2` masking might miss (converters)
-> Only mask the ones `BINARY2` **_cannot_** handle by specifically mentioning it (array types)

Reply:
Just a quick tip here: in GitHub you can use [code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) with the `patch` syntax like so

```patch
diff --git a/astropy/io/votable/tree.py b/astropy/io/votable/tree.py
index 0a1c2fdfd4..9f3880c395 100644
--- a/astropy/io/votable/tree.py
+++ b/astropy/io/votable/tree.py
@@ -3291,14 +3291,23 @@ class TableElement(
                 for row in range(len(array)):
                     array_row = array.data[row]
                     array_mask = array.mask[row]
-
                     if mode == 2:
                         flattened = np.array([np.all(x) for x in array_mask])
                         data.write(converters.bool_to_bitarray(flattened))

                     for i, converter in fields_basic:
                         try:
-                            chunk = converter(array_row[i], array_mask[i])
+                            # BINARY2 cannot handle individual array element masks
+                            converter_type = converter.__self__.__class__
+                            # Delegate converter to handle the mask
+                            delegate_condition = issubclass(
+                                converter_type, converters.Array
+                            )
+                            if mode == 1 or delegate_condition:
+                                chunk = converter(array_row[i], array_mask[i])
+                            else:
+                                # Mask is already handled by BINARY2 behaviour
+                                chunk = converter(array_row[i], None)
                             assert type(chunk) == bytes
                         except Exception as e:
                             vo_reraise(
```